### PR TITLE
Feature/monitor explain api

### PR DESCRIPTION
**Pull request by Tjofil**
Issuer (#751)

Implementation of some basic functionalities listed in issue above. Addition of few integration tests - testing variations of *documents_behind* response field depending on *count_docs* request field, one test demonstrating invalid API invocation (explaining a non document level monitor) and two tests utilizing access roles - another PR in security-analytics is probably required to add an explain role.

*CheckList:*
[] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportMonitorExplainAction.kt [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+                        actionListener,
+                        "monitor",
+                        monitor.id
+                    )
+                ) {
+                    return
+                }
+
+                if (monitor.monitorType == Monitor.MonitorType.DOC_LEVEL_MONITOR) {
+                    log.debug("Doc-level monitor hit.")
```
Comment (petardz) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportMonitorExplainAction.kt:122]: Let's put implementation of this action, for each type of monitor in separate classes

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportMonitorExplainAction.kt [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+                        }
+                    }
+                    actionListener.onResponse(MonitorExplainResponse(monitorId, docDiffAmount))
+                } else {
+                    actionListener.onFailure(
+                        AlertingException("Not allowed to explain this type of monitor!", RestStatus.FORBIDDEN, IllegalStateException())
+                    )
+                }
+            } catch (e: Exception) {
+                actionListener.onFailure(AlertingException.wrap(e))
```
Comment (petardz) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportMonitorExplainAction.kt:188]: log error here

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportMonitorExplainAction.kt [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                        for (i: Int in 0 until count) {
+                            val shard = i.toString()
+
+                            val maxSeqNo: Long = indexUpdatedRunContext[shard].toString().toLong()
+                            val prevSeqNo: Long = indexLastRunContext[shard].toString().toLong()
+
+                            if (!docDiffRequested) {
+                                docDiffAmount += maxSeqNo - prevSeqNo
+                            } else {
+                                val boolQueryBuilder = BoolQueryBuilder()
```
Comment (petardz) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportMonitorExplainAction.kt:160]: nit: extract this block to method

**Comment by dblock** [comment]
Looks stalled, closing. Please feel free to finish and reopen.

[Catch All Triage, attendees [1](https://github.com/dblock), [2](https://github.com/reta), [3](https://github.com/lukas-vlcek/), [4](https://github.com/rafaelma), [5](https://github.com/krisfreedain), [6](https://github.com/gargrajat), [7](https://github.com/andrross)]

**Pull request closed without merging**